File: cpp/libjoynr/joynr-messaging/MutableMessage.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

#include "joynr/MutableMessage.h"

#include <utility>

#include <smrf/MessageSerializer.h>

#include "joynr/IKeychain.h"
#include "joynr/ImmutableMessage.h"
#include "joynr/Message.h"
#include "joynr/Util.h"

namespace joynr
{

MutableMessage::MutableMessage()
        : sender(),
          recipient(),
          _keyChain(nullptr),
          _expiryDate(),
          type(),
          id(util::createUuid()),
          _replyTo(),
          _effort(),
          customHeaders(),
          payload(),
          _localMessage(false),
          _encrypt(false),
          _compress(false)
{
}

std::unique_ptr<ImmutableMessage> MutableMessage::getImmutableMessage() const
{
    smrf::MessageSerializer messageSerializer;

    // propagate flags
    messageSerializer.setCompressed(_compress);

    // explicit headers
    messageSerializer.setSender(sender);
    messageSerializer.setRecipient(recipient);
    messageSerializer.setTtlMs(_expiryDate.toMilliseconds());

    // key-value pair headers
    std::unordered_map<std::string, std::string> keyValuePairHeaders;
    keyValuePairHeaders.insert({Message::HEADER_TYPE(), type});
    keyValuePairHeaders.insert({Message::HEADER_ID(), id});

    if (_replyTo) {
        keyValuePairHeaders.insert({Message::HEADER_REPLY_TO(), *_replyTo});
    }
    if (_effort) {
        keyValuePairHeaders.insert({Message::HEADER_EFFORT(), *_effort});
    }
    keyValuePairHeaders.insert(customHeaders.cbegin(), customHeaders.cend());
    messageSerializer.setHeaders(keyValuePairHeaders);

    smrf::ByteArrayView payloadView(
            reinterpret_cast<smrf::Byte*>(const_cast<char*>(payload.data())), payload.size());
    messageSerializer.setBody(payloadView);

    if (_keyChain) {
        std::string ownerIdStr = _keyChain->getOwnerId();
        smrf::ByteVector ownerIdSignature(ownerIdStr.begin(), ownerIdStr.end());
        auto ownersigningCallback = [ownerIdSignature](const smrf::ByteArrayView&) {
            return ownerIdSignature;
        };
        messageSerializer.setCustomSigningCallback(ownersigningCallback);
    }

    const bool verifyInput = false;
    return std::make_unique<ImmutableMessage>(messageSerializer.serialize(), verifyInput);
}

const std::string& MutableMessage::getRecipient() const
{
    return recipient;
}

const std::unordered_map<std::string, std::string>& MutableMessage::getCustomHeaders() const
{
    return customHeaders;
}

boost::optional<std::string> MutableMessage::getCustomHeader(const std::string& key) const
{
    const std::string lookupKey = Message::CUSTOM_HEADER_PREFIX() + key;
    boost::optional<std::string> value;
    auto it = customHeaders.find(lookupKey);
    if (it != customHeaders.cend()) {
        value = it->second;
    }
    return value;
}

void MutableMessage::setType(std::string&& typeLocal)
{
    this->type = std::move(typeLocal);
}

const std::string& MutableMessage::getType() const
{
    return type;
}

const std::string& MutableMessage::getId() const
{
    return id;
}

void MutableMessage::setReplyTo(const std::string& replyTo)
{
    this->_replyTo = replyTo;
}

const boost::optional<std::string>& MutableMessage::getReplyTo() const
{
    return _replyTo;
}

void MutableMessage::setCustomHeader(const std::string& key, const std::string& value)
{
    customHeaders.insert({Message::CUSTOM_HEADER_PREFIX() + key, value});
}

void MutableMessage::setCustomHeader(std::string&& key, std::string&& value)
{
    customHeaders.insert({Message::CUSTOM_HEADER_PREFIX() + std::move(key), std::move(value)});
}

void MutableMessage::setPrefixedCustomHeaders(
        const std::unordered_map<std::string, std::string>& prefixedCustomHeaders)
{
    customHeaders.insert(prefixedCustomHeaders.cbegin(), prefixedCustomHeaders.cend());
}

void MutableMessage::setPrefixedCustomHeaders(
        std::unordered_map<std::string, std::string>&& prefixedCustomHeaders)
{
    customHeaders.insert(std::make_move_iterator(prefixedCustomHeaders.begin()),
                         std::make_move_iterator(prefixedCustomHeaders.end()));
}

void MutableMessage::setEffort(const std::string& effort)
{
    this->_effort = effort;
}

const boost::optional<std::string>& MutableMessage::getEffort() const
{
    return _effort;
}

void MutableMessage::setPayload(const std::string& payloadLocal)
{
    this->payload = payloadLocal;
}

void MutableMessage::setPayload(std::string&& payloadLocal)
{
    this->payload = std::move(payloadLocal);
}

const std::string& MutableMessage::getPayload() const
{
    return payload;
}

bool MutableMessage::isLocalMessage() const
{
    return _localMessage;
}

void MutableMessage::setLocalMessage(bool localMessage)
{
    this->_localMessage = localMessage;
}

void MutableMessage::setEncrypt(bool encrypt)
{
    this->_encrypt = encrypt;
}

bool MutableMessage::getEncrypt() const
{
    return _encrypt;
}

void MutableMessage::setCompress(bool compress)
{
    this->_compress = compress;
}

bool MutableMessage::getCompress() const
{
    return _compress;
}

void MutableMessage::setEffort(std::string&& effort)
{
    this->_effort = std::move(effort);
}

std::string MutableMessage::toLogMessage() const
{
    return serializer::serializeToJson(*this);
}

void MutableMessage::setType(const std::string& typeLocal)
{
    this->type = typeLocal;
}

TimePoint MutableMessage::getExpiryDate() const
{
    return _expiryDate;
}

void MutableMessage::setExpiryDate(const TimePoint& expiryDate)
{
    this->_expiryDate = expiryDate;
}

void MutableMessage::setRecipient(std::string&& recipientLocal)
{
    this->recipient = std::move(recipientLocal);
}

const std::string& MutableMessage::getSender() const
{
    return sender;
}

void MutableMessage::setRecipient(const std::string& recipientLocal)
{
    this->recipient = recipientLocal;
}

void MutableMessage::setSender(std::string&& senderLocal)
{
    this->sender = std::move(senderLocal);
}

void MutableMessage::setSender(const std::string& senderLocal)
{
    this->sender = senderLocal;
}

void MutableMessage::setKeychain(std::shared_ptr<IKeychain> keyChain)
{
    this->_keyChain = std::move(keyChain);
}

void MutableMessage::setReplyTo(std::string&& replyTo)
{
    this->_replyTo = std::move(replyTo);
}

} // namespace joynr

File: java/messaging/messaging-common/src/main/java/joynr/MutableMessage.java
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package joynr;

import static io.joynr.util.JoynrUtil.createUuidString;

import java.util.HashMap;
import java.util.Map;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;

import io.joynr.smrf.EncodingException;
import io.joynr.smrf.MessageSerializer;
import io.joynr.smrf.MessageSerializerImpl;
import io.joynr.smrf.UnsuppportedVersionException;

/**
 * Represents a joynr message which is about to be transmitted. It will be converted to
 * an {@link ImmutableMessage} before it is passed to a sub-class of
 * {@link io.joynr.messaging.routing.AbstractMessageRouter}.
 */
public class MutableMessage extends Message {
    private String sender;
    private String recipient;
    private long ttlMs;
    private boolean ttlAbsolute;
    private byte[] payload;

    private String id;
    private MessageType type;
    private String replyTo;
    private String effort;
    private Map<String, String> customHeaders = new HashMap<>();

    private transient boolean compressed = false;
    private transient boolean statelessAsync;

    public MutableMessage() {
        id = createUuidString();
    }

    public ImmutableMessage getImmutableMessage() throws SecurityException, EncodingException,
                                                  UnsuppportedVersionException {
        MessageSerializer messageSerializer = new MessageSerializerImpl();

        messageSerializer.setSender(getSender());
        messageSerializer.setRecipient(getRecipient());
        messageSerializer.setTtlMs(getTtlMs());
        messageSerializer.setTtlAbsolute(isTtlAbsolute());
        messageSerializer.setHeaders(createHeader());
        messageSerializer.setBody(getPayload());
        messageSerializer.setCompressed(compressed);

        return new ImmutableMessage(messageSerializer.serialize());
    }

    private Map<String, String> createHeader() {
        Map<String, String> header = new HashMap<>();

        if (customHeaders != null && !customHeaders.isEmpty()) {
            for (Map.Entry<String, String> entry : customHeaders.entrySet()) {
                header.put(CUSTOM_HEADER_PREFIX + entry.getKey(), entry.getValue());
            }
        }
        if (type != null) {
            putIfValueNotNull(Message.HEADER_MSG_TYPE, type.toString(), header);
        }
        putIfValueNotNull(Message.HEADER_ID, id, header);
        putIfValueNotNull(Message.HEADER_REPLY_TO, replyTo, header);
        putIfValueNotNull(Message.HEADER_EFFORT, effort, header);

        return header;
    }

    private void putIfValueNotNull(String key, String value, Map<String, String> destination) {
        if (value != null) {
            destination.put(key, value);
        }
    }

    public String getSender() {
        return sender;
    }

    public void setSender(String sender) {
        this.sender = sender;
    }

    public String getRecipient() {
        return recipient;
    }

    public void setRecipient(String recipient) {
        this.recipient = recipient;
    }

    public long getTtlMs() {
        return ttlMs;
    }

    public void setTtlMs(long ttlMs) {
        this.ttlMs = ttlMs;
    }

    public boolean isTtlAbsolute() {
        return ttlAbsolute;
    }

    public void setTtlAbsolute(boolean ttlAbsolute) {
        this.ttlAbsolute = ttlAbsolute;
    }

    public MessageType getType() {
        return type;
    }

    public void setType(MessageType type) {
        this.type = type;
    }

    public String getId() {
        return id;
    }

    // Findbugs wants us to return a copy. However, this is a getter for the externally provided
    // payload which may change at any time.
    @SuppressFBWarnings("EI_EXPOSE_REP")
    public byte[] getPayload() {
        return payload;
    }

    // Findbugs wants us to copy the parameter. However, the payload is only used during the serialization.
    // Therefore it is okay if it changes.
    @SuppressFBWarnings("EI_EXPOSE_REP")
    public void setPayload(byte[] payload) {
        this.payload = payload;
    }

    public String getReplyTo() {
        return replyTo;
    }

    public void setReplyTo(String replyTo) {
        this.replyTo = replyTo;
    }

    public String getEffort() {
        return effort;
    }

    public void setEffort(String effort) {
        this.effort = effort;
    }

    public Map<String, String> getCustomHeaders() {
        return customHeaders;
    }

    public void setCustomHeaders(Map<String, String> customHeaders) {
        this.customHeaders.putAll(customHeaders);
    }

    public void setCompressed(boolean compressed) {
        this.compressed = compressed;
    }

    public boolean getCompressed() {
        return this.compressed;
    }

    public boolean isStatelessAsync() {
        return statelessAsync;
    }

    public void setStatelessAsync(boolean statelessAsync) {
        this.statelessAsync = statelessAsync;
    }
}
